swxsoc/util/schema.py

```diff
@@ -0,0 +1,1551 @@
+            Guessed CDF Type of the variable
+
+        Returns
+        -------
+        attributes: `OrderedDict`
+            A dict containing `key: value` pairs of derived metadata attributes.
+        """
+        measurement_attributes = OrderedDict()
+
+        # Guess the const CDF Data Type
+        var_data = data[var_name]
+        if not guess_types:
+            if var_name == "time":
+                # Guess the const CDF Data Type
+                (guess_dims, guess_types, guess_elements) = self._types(var_data)
+            elif hasattr(var_data, "value"):
+                # Support NDData use `.value`
+                (guess_dims, guess_types, guess_elements) = self._types(var_data.value)
+            else:
+                # TimeSeries Quantity and Spectra NDCube use `.data`
+                (guess_dims, guess_types, guess_elements) = self._types(var_data.data)
+
+        # Check the Attributes that can be derived
+        var_type = self._get_var_type(var_name, var_data, guess_types[0])
+
+        if var_type in ["data", "support_data", "metadata"]:
+            # Derive Attributes Specific to VAR_TYPE
+            derived_attributes = filter(
+                lambda attr_info: attr_info[0]
+                in self.variable_attribute_schema[var_type]
+                and attr_info[1]["derived"],
+                self.variable_attribute_schema["attribute_key"].items(),
+            )
+            for attr_name, attr_schema in derived_attributes:
+                derivation_fn = getattr(self, attr_schema["derivation_fn"])
+                measurement_attributes[attr_name] = derivation_fn(
+                    var_name, var_data, guess_types[0]
+                )
+        else:
+            warn_user(
+                f"Variable {var_name} has unrecognizable VAR_TYPE ({var_type}). Cannot Derive Metadata for Variable."
+            )
+
+        # Derive Attributes Specific to `spectra` Data
+        if hasattr(var_data, "wcs") and getattr(var_data, "wcs") is not None:
+            spectra_attributes = self._derive_spectra_attributes(var_data)
+            measurement_attributes.update(spectra_attributes)
+
+        return measurement_attributes
+
+    def derive_time_attributes(self, data) -> OrderedDict:
+        """
+        Function to derive metadata for the time measurement.
+
+        Parameters
+        ----------
+        data : `swxsoc.swxdata.SWXData`
+            An instance of `SWXData` to derive metadata from.
+
+        Returns
+        -------
+        attributes : `OrderedDict`
+            A dict containing `key: value` pairs of time metadata attributes.
+        """
+
+        # Get the Variable Data
+        var_data = data["time"]
+        (guess_dims, guess_types, guess_elements) = self._types(var_data)
+
+        time_attributes = self.derive_measurement_attributes(
+            data, "time", guess_types=guess_types
+        )
+        # Check the Attributes that can be derived
```

Comment:
There is a subset of attributes derived for `time` that are hardcoded and not determined based on the schema. This function would need to be overridden if you wan't to have a different subset of time-specific attributes.